From e8bc63033a2b7b319397612a7807d7618f0782b1 Mon Sep 17 00:00:00 2001
From: Michael Cook <code@mdcook.net>
Date: Mon, 1 Jan 2018 14:27:51 -0800
Subject: [PATCH 050/249] Initial commit of device schema validation

- currently only partial device validation, will add each device
  property until all properties are validated and then enforce
  property validation
- update gemfile with gems used for validation testing

Change-Id: Ied96ca0444282c77db758a0a6eebd7d75c6e219d
---
 Gemfile            |  3 ++
 Gemfile.lock       |  3 ++
 test/schema-06.yml | 95 ++++++++++++++++++++++++++++++++++++++++++++++
 test/validate.rb   | 40 +++++++++++++++++++
 4 files changed, 141 insertions(+)
 create mode 100644 test/schema-06.yml
 create mode 100755 test/validate.rb

diff --git a/Gemfile b/Gemfile
index 706d7ee..10e1cc5 100644
--- a/Gemfile
+++ b/Gemfile
@@ -4,6 +4,9 @@ source "https://rubygems.org"
 gem "jekyll", "3.6.2"
 gem "therubyracer", "0.12.3"
 
+group :test do
+   gem 'json-schema', '= 2.8.0'
+end
 
 group :jekyll_plugins do
    gem 'activesupport', '= 4.2.9'
diff --git a/Gemfile.lock b/Gemfile.lock
index 91ba7ad..b1d0dbb 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -76,6 +76,8 @@ GEM
       gemoji (~> 3.0)
       html-pipeline (~> 2.2)
       jekyll (>= 3.0)
+    json-schema (2.8.0)
+      addressable (>= 2.4)
     kramdown (1.14.0)
     libv8 (3.16.14.19)
     liquid (4.0.0)
@@ -142,6 +144,7 @@ DEPENDENCIES
   jekyll-sitemap (= 1.1.1)
   jekyll-swiss (= 0.4.0)
   jemoji (= 0.8.1)
+  json-schema (= 2.8.0)
   kramdown (= 1.14.0)
   liquid (= 4.0.0)
   listen (= 3.0.6)
diff --git a/test/schema-06.yml b/test/schema-06.yml
new file mode 100644
index 0000000..218173a
--- /dev/null
+++ b/test/schema-06.yml
@@ -0,0 +1,95 @@
+---
+"$schema": http://json-schema.org/draft-06/schema#
+definitions: {}
+properties:
+  architecture:
+    "$id": "/properties/architecture"
+    enum:
+    - arm
+    - arm64
+    - x86
+    - x86_64
+  channels:
+    "$id": "/properties/channels"
+    items:
+      "$id": "/properties/channels/items"
+      enum:
+      - discontinued
+      - weekly
+    type: array
+  codename:
+    "$id": "/properties/codename"
+    type: string
+  type:
+    "$id": "/properties/type"
+    enum:
+    - Handheld game console
+    - Set top box
+    - phablet
+    - phone
+    - phone (slider)
+    - tablet
+  vendor:
+    "$id": "/properties/vendor"
+    enum:
+    - ARK
+    - Asus
+    - BQ
+    - Fairphone
+    - Google
+    - HTC
+    - Huawei
+    - LG
+    - LeEco
+    - Lenovo
+    - Motorola
+    - Nextbit
+    - Nubia
+    - Nvidia
+    - OPPO
+    - OnePlus
+    - Samsung
+    - Sony
+    - Wileyfox
+    - Wingtech
+    - Xiaomi
+    - YU
+    - ZTE
+    - ZUK
+  vendor_short:
+    "$id": "/properties/vendor_short"
+    enum:
+    - ark
+    - asus
+    - bq
+    - fairphone
+    - google
+    - htc
+    - huawei
+    - leeco
+    - lenovo
+    - lge
+    - moto
+    - motorola
+    - nextbit
+    - nubia
+    - nvidia
+    - oneplus
+    - oppo
+    - samsung
+    - sony
+    - wileyfox
+    - wingtech
+    - xiaomi
+    - yu
+    - zte
+    - zuk
+required:
+- architecture
+- channels
+- codename
+- type
+- vendor
+- vendor_short
+title: phone
+type: object
diff --git a/test/validate.rb b/test/validate.rb
new file mode 100755
index 0000000..68b1b94
--- /dev/null
+++ b/test/validate.rb
@@ -0,0 +1,40 @@
+#!/usr/bin/env ruby
+
+require 'json'
+require 'json-schema'
+require 'yaml'
+
+
+def json_to_yaml(json)
+  JSON.parse(json).to_yaml
+end
+
+def yaml_to_json(yaml)
+   YAML.load_file(yaml).to_json
+end
+
+schema_path = File.expand_path('schema-06.yml', __dir__)
+schema = yaml_to_json(schema_path)
+
+sample_path = File.expand_path('../device_sample/sample.yml', __dir__)
+sample = yaml_to_json(sample_path)
+
+begin
+  JSON::Validator.validate!(schema, sample, :validate_schema => true)
+rescue JSON::Schema::ValidationError => e
+  puts sample_path + ': ' + e.message
+end
+
+device_dir = File.expand_path('../_data/devices/', __dir__) + '/'
+Dir.entries(device_dir).each do |filename|
+  device_path = device_dir + filename
+  if File.file?(device_path)
+    device = yaml_to_json(device_path)
+
+    begin
+      JSON::Validator.validate!(schema, device, :validate_schema => true)
+    rescue JSON::Schema::ValidationError => e
+      puts device_path + ': ' + e.message
+    end
+  end
+end
-- 
2.17.1

